typed-racket-lib/typed-racket/private/type-contract.rkt

```diff
@@ -1435,116 +1435,88 @@
-
-(define merge-vector/sc*
-  (make-merge-overlapping-function
-    (λ (this-sc)
-      (match this-sc
-       [(or (immutable-vector/sc: this-TS)
-            (mutable-vector/sc: this-TS)
-            (vector/sc: this-TS))
-        (cons (vector/sc this-TS)
-              (λ (that-sc)
-                (match that-sc
-                 [(or (immutable-vector/sc: that-TS)
-                      (mutable-vector/sc: that-TS)
-                      (vector/sc: that-TS))
-                  (equal? this-TS that-TS)]
-                 [_ #false])))]
-       [_ #false]))))
-
-(define merge-vectorof/sc*
-  (make-merge-overlapping-function
-    (λ (this-sc)
-      (match this-sc
-       [(or (immutable-vectorof/sc: this-T)
-            (mutable-vectorof/sc: this-T)
-            (vectorof/sc: this-T))
-        (cons (vectorof/sc this-T)
-              (λ (that-sc)
-                (match that-sc
-                 [(or (immutable-vectorof/sc: that-T)
-                      (mutable-vectorof/sc: that-T)
-                      (vectorof/sc: that-T))
-                  (equal? this-T that-T)]
-                 [_ #false])))]
-       [_ #false]))))
+(define (hash-type? ty)
+  (match ty
+   [(or (Mutable-HashTable: _ _)
+        (Mutable-HashTableTop:)
+        (Immutable-HashTable: _ _)
+        (Weak-HashTable: _ _)
+        (Weak-HashTableTop:))
+    #true]
+   [_ #false]))
+
+(define (vector-type? ty)
+  (match ty
+   [(or (Immutable-Vector: _)
+        (Mutable-Vector: _)
+        (Mutable-VectorTop:))
+    #true]
+   [_ #false]))
+
+(define (heterogeneous-vector-type? ty)
+  (match ty
+   [(or (Immutable-HeterogeneousVector: _)
+        (Mutable-HeterogeneousVector: _))
+    #true]
+   [_ #false]))
+
+(define (merge-hash-types tys t->sc top-sc)
+  (match tys
+   [(HashTableTop:)
+    (list top-sc)]
+   [(list-no-order
+      (and t0 (Immutable-HashTable: k0 v0))
+      (and t1 (Mutable-HashTable: k1 v1))
+      (and t2 (Weak-HashTable: k2 v2)))
+    (define =01 (and (equal? k0 k1) (equal? v0 v1)))
+    (define =02 (and (equal? k0 k2) (equal? v0 v2)))
+    (define =12 (and (equal? k1 k2) (equal? v1 v2)))
+    (cond
+     [(and =01 =12)
+      (list (hash/sc (t->sc k0) (t->sc v0)))]
+     [=12
+      (list (hash/sc k1 v1) (t->sc t0))]
+     [=01
+      (list (hash/sc k0 v0) (t->sc t2))]
+     [=02
+      (list (hash/sc k0 v1) (t->sc t1))]
+     [else
+      (map t->sc tys)])]
+   [(list-no-order
+      (Immutable-HashTable: k0 v0)
+      (Mutable-HashTable: k1 v1))
+    #:when (and (equal? k0 k1) (equal? v0 v1))
+    (list (hash/sc k0 v1))]
+   [(list-no-order
+      (Immutable-HashTable: k0 v0)
+      (Weak-HashTable: k1 v1))
+    #:when (and (equal? k0 k1) (equal? v0 v1))
+    (list (hash/sc k0 v1))]
+   [(list-no-order
+      (Mutable-HashTable: k0 v0)
+      (Weak-HashTable: k1 v1))
+    #:when (and (equal? k0 k1) (equal? v0 v1))
+    (list (hash/sc k0 v1))]
+   [_
+    (map t->sc tys)]))
+
+(define (merge-vector-types tys t->sc top-sc)
```

Comment:
I am not sure I am completely following the reasoning about all `merge-*` operations here. It looks to me that they only generate a merged static contract when the input `tys` is a list of two elements that meet some conditions.  

let's say that `tys` is `(list (Immutable-Vectorof Natural) (Immutable-Vectorof String) (Mutable-Vectorof Natural))`. Should we merge the static contract for  `(Immutable-Vectorof Natural)` and `(Mutable-Vectorof Natural))`.

Reply:
That's a bug. These functions need editing.

They'd better be recursive too. If we change your example to `((IV Natural) (IV String) (MV Natural) (MV String))` the end result should be `((v natural) (v string))`

Reply:
hooray for `group-by`